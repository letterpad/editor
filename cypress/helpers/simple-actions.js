require("cypress-xpath");

export async function repeatKey(key, times) {
  for (let i = 0; i < times; i++) {
    await cy.focused().type(key);
  }
}

export async function clearEditor() {
  return new Promise((resolve, reject) => {
    cy.window().then(win => {
      win.__letterpadEditor.moveToRangeOfDocument();
      cy.focused()
        .type("{backspace}", { force: true })
        .then(resolve);
    });
  });
}

export async function clickXPath(xPath) {
  const ele = cy.xpath(xPath);
  if (ele.length > 0) {
    ele.click({ multiple: true });
  } else {
    ele.click();
  }
}
