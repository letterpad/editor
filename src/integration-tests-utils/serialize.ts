import { ElementHandle } from "puppeteer";
import cheerio from "cheerio";

export type EditorHandle = ElementHandle<Element>;

export async function getHtmlContents(handle: EditorHandle): Promise<string> {
  const context = await handle.executionContext();
  const html = await context.evaluate(node => node.innerHTML, handle);

  const cHtml = cheerio.load(html);
  const allNodes = cHtml("*");
  for (let i = 0; i < allNodes.length; i++) {
    const node = allNodes[i];
    for (let attr in node.attribs) {
      if (attr.startsWith("data-")) {
        allNodes.removeAttr(attr);
      }
    }
  }

  return cHtml.html();
}

export async function textContent(handle: EditorHandle): Promise<any> {
  const textContent = await handle.getProperty("textContent");
  return textContent.jsonValue();
}
