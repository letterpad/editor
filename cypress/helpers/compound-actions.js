// import { cy } from "./serialize";
import { clearEditor, repeatKey, clickXPath } from "./simple-actions";

export async function applyEditorFeatureToLine(xPath) {
  return new Promise(resolve => {
    cy.get('div[contenteditable="true"]').then(() => {
      cy.focused()
        .type("These is a sample line of text{selectAll}")
        .then(() => {
          clickXPath(xPath);
          resolve();
        });
    });
  });
}

export async function applyEditorFeatureToSampleText(xPath) {
  return new Promise(resolve => {
    cy.get('div[contenteditable="true"]').then(() => {
      cy.focused()
        .type("This is a sample text{selectAll}")
        .then(() => {
          clickXPath(xPath);
          resolve();
        });
    });
  });
  // await cy.focused().type("{selectall}{rightarrow}{rightarrow}");
  // // .setSelection("sample");
  // await repeatKey("{leftarrow}", 5); // move 5 left
  // await cy.focused().type("{shift}", { release: false });
  // await repeatKey("{shift}{leftarrow}", 6); // move 6 left
  // await cy.focused().type("{shift}", { release: true });
  // // await cy.focused().type("{selectAll}");
  // await clickXPath(xPath);
}

export async function fillSampleList(cy) {
  await clearEditor(cy);
  await page.keyboard.type("- Item 1");
  await page.keyboard.press("Enter");
  await page.keyboard.type("Item 2");
  await page.keyboard.press("Enter");
  await page.keyboard.press("Tab");
  await page.keyboard.type("Item 2.1");
  await page.keyboard.press("Enter");
  await page.keyboard.type("Item 2.2");
  await page.keyboard.press("Enter");
  await page.keyboard.press("Tab");
  await page.keyboard.type("Item 2.2.1");
  await page.keyboard.press("Enter");
  await page.keyboard.type("Item 2.2.2");
  await page.keyboard.press("Enter");
  await page.keyboard.type("Item 2.2.3");
}
