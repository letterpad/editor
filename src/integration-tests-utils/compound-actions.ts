import { EditorHandle } from "./serialize";
import { clearEditor, repeatKey } from "./simple-actions";

export async function applyEditorFeatureToSampleText(
  editorHandle: EditorHandle,
  xPath: string
) {
  await clearEditor(editorHandle);
  await page.keyboard.type("This is a sample text");
  const textHandlers = await page.$x("//span[contains(text(), 'sample')]");
  if (textHandlers.length < 1) {
    throw new Error("Expected at least 1 matching textHandler");
  }

  await repeatKey("ArrowLeft", 5); // move 5 left
  await page.keyboard.down("Shift");
  await repeatKey("ArrowLeft", 6); // move 6 left
  await page.keyboard.up("Shift");
  await page.keyboard.down("Meta");
  await page.keyboard.press("b");
  await page.keyboard.up("Meta");
  await page.waitFor(50);

  const featureHandlers = await page.$x(xPath);
  if (featureHandlers.length < 1) {
    throw new Error(`Feature Button xpath="${xPath}" not found`);
  }

  featureHandlers[0].click();
  await page.waitFor(50);
}
