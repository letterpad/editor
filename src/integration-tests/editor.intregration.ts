import {
  EditorHandle,
  getHtmlContents
} from "../integration-tests-utils/serialize";
import { clearEditor } from "../integration-tests-utils/simple-actions";
import { applyEditorFeatureToSampleText } from "../integration-tests-utils/compound-actions";

describe("functionalities", () => {
  let editorHandle: EditorHandle;

  beforeAll(async () => {
    await page.goto("http://localhost:4343");
    await page.waitForXPath("/html/body/div");
    const handle = await page.$('div[contenteditable="true"]');
    editorHandle = handle!;
    await editorHandle.focus();
  });

  test("clearing editor", async () => {
    await clearEditor(editorHandle);
    expect(await getHtmlContents(editorHandle!)).toMatchSnapshot();
  });

  test("adding some text", async () => {
    await clearEditor(editorHandle);
    await page.keyboard.type("Hello World!");
    expect(await getHtmlContents(editorHandle!)).toMatchSnapshot();
  });

  test("bold", async () => {
    await applyEditorFeatureToSampleText(
      editorHandle,
      "//span[contains(text(), 'format_bold')]"
    );
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });
});
