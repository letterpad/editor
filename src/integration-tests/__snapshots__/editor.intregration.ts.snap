// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features adding some text 1`] = `"<html><head></head><body><div style=\\"position: relative;\\"><span><span><span>Hello World!</span></span></span></div></body></html>"`;

exports[`features bold 1`] = `"<html><head></head><body><div style=\\"position: relative;\\"><span><span><span>This is a </span></span><span><strong><span>sample</span></strong></span><span><span> text</span></span></span></div></body></html>"`;

exports[`features clearing editor 1`] = `"<html><head></head><body><div style=\\"position: relative;\\"><span><span><span><span contenteditable=\\"false\\" style=\\"pointer-events: none; display: inline-block; width: 0px; max-width: 100%; white-space: nowrap; opacity: 0.333;\\">Compose a story..</span><span>&#xFEFF;<br></span></span></span></span></div></body></html>"`;

exports[`features italics 1`] = `"<html><head></head><body><div style=\\"position: relative;\\"><span><span><span>This is a </span></span><span><em><span>sample</span></em></span><span><span> text</span></span></span></div></body></html>"`;

exports[`features underline 1`] = `"<html><head></head><body><div style=\\"position: relative;\\"><span><span><span>This is a </span></span><span><u><span>sample</span></u></span><span><span> text</span></span></span></div></body></html>"`;
